### ISSUE TITLE
feat: HTML template support

### ISSUE THREAD

**Issue by paurakhsharma** [VeryGoodOpenSource/dart_frog#90]
**Description**
As a user, I want to be able to respond with an HTML template. 

If we don't want to implement it in this core project. Would love to see separate project to support it as a `plugin` either by VG team themselves or by the community.

**Comment by felangel** [comment]
Hi @paurakhsharma 👋 
Thanks for opening an issue!

Can you please provide a bit more context regarding the expectation? Currently, you can create an endpoint that responds with html like:

```dart
import 'dart:io';

import 'package:dart_frog/dart_frog.dart';

Response onRequest(RequestContext context) {
  return Response(
    body: '<p>Hello World</p>',
    headers: {HttpHeaders.contentTypeHeader: ContentType.html.value},
  );
}
```

**Comment by paurakhsharma** [comment]
That is awesome. I can see we can use `String interpolation` for adding some values from the variable.
But other things like looping or conditional rendering might be difficult.

I was thinking something like.

```dart
import 'dart:io';

import 'package:dart_frog/dart_frog.dart';

Response onRequest(RequestContext context) {
  return RenderTemplate(
            'index.html',
             params: {
                 "hello": "world",
                 "bestLanguages": ["Dart", "Dart"],
                 "isDartFrogAwesome": true,
              }
        );
}
```

And in `index.html`

```html
<p>Hello {{hello}}</p>

<div>
   {% for language in bestLanguages %}
        <li>{{ language|toUpperCase }}</li>
    {% endfor %}
</div>

{% if isDartFrogAwesome %}
    <p>Dart Frog is awesome 🚀</p>
{% endif %}
```

**Comment by felangel** [comment]
@paurakhsharma thanks for clarifying! I definitely think this can be supported via a separate package. I’ll discuss with the team and update the ROADMAP accordingly, thanks again 🙏

**Comment by paurakhsharma** [comment]
Amazing! Thank you so much 😊

**Comment by AlvaroVasconcelos** [comment]
What do you think of OverReact to assemble the html?
```dart 
render() {
  return (Dom.div()..className = 'container')(
    Dom.h1()('Click the button!'),
    (Dom.button()
      ..id = 'main_button'
      ..onClick = _handleClick
    )('Click me')
  );
}
```

**Comment by Mastersam07** [comment]
Hi, @felangel is it possible to return an HTML file directly in the response?

**Comment by ushieru** [comment]
@Mastersam07 Yes, you can.

```dart
// routes/index.dart

import 'dart:io';

import 'package:dart_frog/dart_frog.dart';
import 'package:path/path.dart' as path;

Response onRequest(RequestContext context) {
  final file = File(path.join(Directory.current.path, 'public', 'index.html'));
  if (!file.existsSync()) {
    return Response(body: 'Index Not found');
  }
  final indexHtml = file.readAsStringSync();
  return Response(body: indexHtml, headers: {'Content-Type': 'text/html'});
}

```

**Comment by felangel** [comment]
If you just want to respond with a static html file you can just [serve the static asset](https://dartfrog.vgv.dev/docs/basics/serving-static-files)

**Comment by Mastersam07** [comment]
Thanks, @ushieru @felangel. I already found a solution to the issue.

**Comment by AlvaroVasconcelos** [comment]
What do you guys think of [jasper](https://github.com/schultek/jaspr)?

**Comment by Eghosa-Osayande** [comment]
Hello @paurakhsharma @felangel 

I recently wrote an article that illustrates how to use HTML templates in a dart frog application using the jinja package.
https://yande.hashnode.dev/jinja-html-templates-with-dart-frog

A Quick Explanation:

pubspec.yaml
```
dependencies:
    ...
    jinja: ^0.5.0

```

routes/index.dart
```dart
import 'package:dart_frog/dart_frog.dart';
import 'package:jinja/jinja.dart';
import 'package:jinja/loaders.dart';

Response onRequest(RequestContext context) {
  final environment = Environment(
    loader: FileSystemLoader(
      paths: [
        'templates'
      ],
    ),
  );

  final template = environment.getTemplate(
    'base.html',
  );

  final tutorials = [
    'Jinja Templates Tutorial',
    'Flutter Tutorials',
    'Dart Fundamentals',
  ];

  return Response(
    body: template.render({
      'title': 'Home',
      'tutorials': tutorials,
    }),
    headers: {
      'Content-Type': 'text/html'
    },
  );
}
``` 

templates/base.html
```html
<html lang="en">

<head>

  <meta charset="utf-8">

  <title>{{ title }}</title>

</head>

<body>

  <h1>Welcome to {{ title }}!</h1>

  {% for tutorial in tutorials %}

  <ul>

    <li>

      {{ tutorial }}</em>

    </li>

  </ul>

  {% endfor %}

</body>

</html>
``` 

Output

<img width="477" alt="Screenshot 2023-08-22 at 17 21 39" src="https://github.com/VeryGoodOpenSource/dart_frog/assets/69315505/37b06172-1b33-438b-b039-779478aeeb4f">

I hope this helps.

**Comment by Mastersam07** [comment]
Nice read @Eghosa-Osayande 

I also resorted to using jinja templating given that i have background in django and used jinja template in the past. I forgot to mention in my last update on this issue.

**Comment by moshe5745** [comment]
@paurakhsharma Nice proposal.

It seems that `jinja`(like @Eghosa-Osayande propoused) is exactly what we need.

**Comment by moshe5745** [comment]
You can try another package if you want static type
https://github.com/xvrh/html_template

**Comment by tomarra** [comment]
Hi everyone 👋Thanks for the discussion on this topic. It’s a ticket that has been open for a while now so it’s great to see some long running contributors.

Overall, the VGV team has not run into a use case that requires this functionality so we haven’t prioritized it. It looks like, based off of @Eghosa-Osayande and @moshe5745’s research/comments, that there are now some community packages that focus on the templating use case and can be used in conjunction with Dart Frog. This feels like the better approach as it keeps the Dart Frog framework more generic and allows for people to choose a templating framework that fits their needs the best.

One thing that may be worth doing is writing up this use of Dart Frog as either an example application or in our docs so others can find this information more easily as well. 

I’m going to close this ticket out as won’t fix given this but if there is someone who does want to work on an example or documentation on this please feel free to open up a new issue for that work so we can get it scoped properly.

**Comment by moshe5745** [comment]
I just made some base example for SSR with dart frog + htmlx templates + htmx + tailwind+ daisyUi
Including hot reload for the: templates + tailwind and browser sync for refresh the browser.

This project included npm.
At first I wanted to avoid it with a standalone tailwind. But I switched to npm because daisy is third-party plugin so the standalone tailwind didn't support it.

@tomarra Is this kind of example is suitable?
